mlflow/models/evaluation/default_evaluator.py

```diff
@@ -647,12 +649,19 @@ def _log_model_explainability(self):
                 )
             else:
                 shap_values = explainer(sampled_X)
+            # We want shap plots to use self.feature_names, but we can't use always use
+            # self.feature_names when computing shap_values, as they may differ from the
+            # names used during training (e.g. when using np.ndarrays). Instead, we rename
+            # shap_values.feature_names here before creating the plots.
+            shap_values.feature_names = [
+                generate_feature_name_if_not_string(n) for n in self.feature_names
+            ]
```

Comment:
I successfully ran every test from `test_default_evaluator.py` and don't seem to have any issues creating the explainability plots. So this at least tests the default explainer and the kernel explainer. Do you happen to remember what issues you were getting by changing `shap_values.feature_names` here? It would be good to have a test case that highlights this problem.

Reply:
> Do you happen to remember what issues you were getting by changing shap_values.feature_names here?

Previously I didn't test against approach of setting `shap_values.feature_names` ,
but I hit issue `feature_name set for explainer is not used, but feature_name in dataframe is used`. Shap code is not maintained well and some edge cases might not work well, so I suggest testing feature names with following cases:

Test on Sklearn logistic model ,use default explainer (i.e. linear explainer)
Test on xgboost classifier, use default explainer (i.e. tree explainer)
Test on any model, use specific explainer (via setting evaluator config `explainability_algorithm`): 'exact', 'permutation', 'partition', 'kernel'

If they all work fine, then it's good.

Reply:
@WeichenXu123 `Explanation.feature_names` comes with a proper setter function to alter the underlying slicer array `Explanation._s`. This setter function is used twice in `shap/_explanation.py`. For example,

```
def _numpy_func(self, fname, **kwargs):
        """ Apply a numpy-style function to this Explanation.
        """
        new_self = copy.copy(self)
        ...
        ...
        ...
            new_self.feature_names = np.array(list(new_values.keys()))
```

The other use in their source code can be found in `Explanation.__getitem__`.

I couldn't find any official API information other than: https://shap.readthedocs.io/en/latest/generated/shap.Explanation.html?highlight=feature_names#shap.Explanation.feature_names - but this doesn't show `feature_names` being set explicitly by the user. There are other elements in this slicer array that don't get setter methods though.

Do you think it is safe to use the setter for `shap_values.feature_names`? If so, then I can test those various explainability_algorithms. I can't foresee there being any issues as this is a postprocessing step that executes after the algorithm has been run - but I can double-check and add explicit test cases to `test_default_evaluator.py`.

Reply:
OK. Seems shap doc does not explicitly said we can set the `shap_values.feature_names`, it seems unsafe to do so.

So I suggest we forbid EvaluationDataset input with "unlabeled pandas dataframe", WDYT ?

Reply:
@WeichenXu123 I disagree about forbidding unlabeled dataframes, as we get the exact same warning when using NumPy arrays. For example, the following block evaluates on NumPy arrays but generates the same warnings as if `X_test` and `y_test` were combined into a single unlabeled dataframe.

```
import mlflow

import pandas as pd
from sklearn import datasets
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# prepare train and test data
iris = datasets.load_iris()
X = iris.data
y = iris.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

with mlflow.start_run():
    logistic = LogisticRegression(C=1.0)
    logistic.fit(X_train, y_train)
    mod = mlflow.sklearn.log_model(logistic, artifact_path="mymodel")

    results = mlflow.evaluate(
        model=mod.model_uri,
        model_type="classifier",
        data=X_test,
        targets=y_test,
        evaluator_config={
            "log_model_explainability": True,
        },
        evaluators="default",
    )
```

If we cannot change `shap_values.feature_names` using the built-in setter, then we probably can't avoid this warning for now (well, at least I don't know of a way that doesn't end up breaking `test_pipeline_model_kernel_explainer_on_categorical_features`). 

If we cannot resolve this issue, then I can just strip out the sklearn warning fix, and just leave in the renaming/docstring for `explainability_ignore_exceptions`.